src/Control/Carrier/Accum/Church.hs

```diff
@@ -0,0 +1,133 @@
+, module Control.Effect.Accum
+) where
+
+import Control.Algebra
+import Control.Applicative (Alternative(..))
+import Control.Effect.Accum
+import Control.Monad (MonadPlus(..))
+import Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+
+-- | Run an 'Accum' effect with a 'Monoid'al log, applying a continuation to the final log and result.
+--
+-- @
+-- 'runAccum' k w0 ('pure' a) = k 'w0' a
+-- @
+-- @
+-- 'runAccum' k w0 ('add' w) = k (w0 <> w) ()
+-- @
+-- @
+-- 'runAccum' k w0 ('add' w >> 'look') = k (w0 <> w) (w0 <> w)
+-- @
+--
+-- @since 1.1.1.0
+runAccum :: (w -> a -> m b) -> w -> AccumC w m a -> m b
+runAccum k w ma = runAccumC ma k w
+{-# INLINE runAccum #-}
+
+-- | Run an 'Accum' effect (typically with a 'Monoid'al log),
+--   producing the final log and discarding the result value.
+--
+-- @
+-- 'execAccum' = 'runAccum' ('const' '.' 'pure')
+-- @
+--
+-- @since 1.1.1.0
+execAccum :: Applicative m => w -> AccumC w m a -> m w
+execAccum = runAccum (const . pure)
+{-# INLINE execAccum #-}
+
+-- | Run an 'Accum' effect (typically with a 'Monoid'al log),
+--   producing the result value and discarding the final log.
+--
+-- @
+-- 'evalAccum' = 'runAccum' ('const' '.' 'pure')
+-- @
+--
+-- @since 1.1.1.0
+evalAccum :: Applicative m => w -> AccumC w m a -> m a
+evalAccum = runAccum $ const pure
+{-# INLINE evalAccum #-}
+
+-- | @since 1.1.1.0
+newtype AccumC w m a = AccumC { runAccumC :: forall r . (w -> a -> m r) -> w -> m r }
+
+instance Monoid w => MonadTrans (AccumC w) where
+  lift ma = AccumC $ \k _ -> ma >>= k mempty
+  {-# INLINE lift #-}
+
+instance Functor (AccumC w m) where
+  fmap f ma = AccumC $ \k w -> runAccumC ma (\w a -> k w $ f a) w
+  {-# INLINE fmap #-}
+
+instance Monoid w => Applicative (AccumC w m) where
+  pure a = AccumC $ \k _ -> k mempty a
+  {-# INLINE pure #-}
+
+  mf <*> ma = AccumC $ \k w ->
+    runAccumC mf (\w' f -> runAccumC ma (\w'' a -> k (w' <> w'') $ f a) (w <> w')) w
+  {-# INLINE (<*>) #-}
+
+instance (Alternative m, Monad m, Monoid w) => Alternative (AccumC w m) where
+  empty = lift empty
+  {-# INLINE empty #-}
+
+  ma1 <|> ma2 = AccumC $ \k w -> runAccumC ma1 k w <|> runAccumC ma2 k w
+  {-# INLINE (<|>) #-}
+
+instance (Monad m, Monoid w) => Monad (AccumC w m) where
+  ma >>= f = AccumC $ \k w -> runAccumC ma (\w' a -> runAccumC (f a) (\w'' -> k $ w' <> w'') (w <> w')) w
+  {-# INLINE (>>=) #-}
+
+instance (MonadPlus m, Monoid w) => MonadPlus (AccumC w m) where
+  mzero = lift mzero
+  {-# INLINE mzero #-}
+
+  ma1 `mplus` ma2 = AccumC $ \k w -> runAccumC ma1 k w `mplus` runAccumC ma2 k w
+  {-# INLINE mplus #-}
+
+instance (MonadFail m, Monoid w) => MonadFail (AccumC w m) where
+  fail msg = AccumC $ const $ const $ Fail.fail msg
+  {-# INLINE fail #-}
+
+instance (MonadFix m, Monoid w) => MonadFix (AccumC w m) where
+  mfix ma = AccumC $ \k w -> fmap fst $ mfix $ \(_, a) -> runAccumC (ma a) (\_ a' -> ( , a') <$> k w a') w
```

Comment:
```suggestion
  mfix ma = AccumC $ \ k w -> fmap fst $ mfix $ \ ~(_, a) -> runAccumC (ma a) (\ _ a' -> ( , a') <$> k w a') w
```

That might do it ☝🏻 As with `fix`, if the function passed to `mfix` strictly evaluates its argument, it diverges.